### DISCUSSION TITLE
CPU overheat

### DISCUSSION THREAD

**Discussion by amit-z** [ultralytics/yolov5#8313] [Q&A]
Hey,
I am using yolov5 detection with my own trained weights, in a ROS node, using the CPU.
After a couple of seconds, the CPU overheats, which makes the detection much slower.

Does anybody have any tips on how to reduce the resources the detection uses?

**Comment by glenn-jocher** [comment]
@amit-z 👋 Hello! Thanks for asking about **inference speed issues**. PyTorch Hub speeds will vary by hardware, software, model, inference settings, etc. Our default example in Colab with a V100 looks like this:

<img width="1462" alt="Screen Shot 2022-05-03 at 10 20 39 AM" src="https://user-images.githubusercontent.com/26833433/166505744-b7fe9b9a-d2b3-4e95-9441-fa540e0804ef.png">


YOLOv5 🚀 can be run on CPU (i.e. `--device cpu`, slow) or GPU if available (i.e. `--device 0`, faster). You can determine your inference device by viewing the YOLOv5 console output:

### detect.py inference
```bash
python detect.py --weights yolov5s.pt --img 640 --conf 0.25 --source data/images/
```

<img width="1039" alt="Screen Shot 2022-05-03 at 2 48 42 PM" src="https://user-images.githubusercontent.com/26833433/166572240-ae598b34-1084-40f0-a19b-bf00fd3f9050.png">


### YOLOv5 PyTorch [Hub](https://github.com/ultralytics/yolov5/issues/36) inference
```python
import torch

# Model
model = torch.hub.load('ultralytics/yolov5', 'yolov5s')

# Images
dir = 'https://ultralytics.com/images/'
imgs = [dir + f for f in ('zidane.jpg', 'bus.jpg')]  # batch of images

# Inference
results = model(imgs)
results.print()  # or .show(), .save()
# Speed: 631.5ms pre-process, 19.2ms inference, 1.6ms NMS per image at shape (2, 3, 640, 640)
```

### Increase Speeds

If you would like to increase your inference speed some options are:

- Use batched inference with [YOLOv5 PyTorch Hub](https://github.com/ultralytics/yolov5/issues/36)
- Reduce `--img-size`, i.e. 1280 -> 640 -> 320
- Reduce model size, i.e. YOLOv5x -> YOLOv5l -> YOLOv5m -> YOLOv5s -> YOLOv5n
- Use half precision FP16 inference with `python detect.py --half` and `python val.py --half`
- Use a faster GPUs, i.e.: P100 -> V100 -> A100
- [Export](https://github.com/ultralytics/yolov5/issues/251) to ONNX or OpenVINO for up to 3x CPU speedup ([CPU Benchmarks](https://github.com/ultralytics/yolov5/pull/6613))
- [Export](https://github.com/ultralytics/yolov5/issues/251) to TensorRT for up to 5x GPU speedup ([GPU Benchmarks](https://github.com/ultralytics/yolov5/pull/6963))
- Use a free GPU backends with up to 16GB of CUDA memory: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

Good luck 🍀 and let us know if you have any other questions!
